perf_benchmark/benchmark_main.cc

```diff
@@ -0,0 +1,497 @@
+      state.SkipWithError(absl::StrFormat(
+          "Actual layer is not as expected. Actual layers: %d, Expected layers: %d",
+          actual_layers,
+          layers).c_str());
+    }
+  }
+
+}
+
+static void BM_NestedProtoPayloadFromJsonNonStreaming(::benchmark::State& state) {
+  BM_NestedPayloadFromJson(state,
+                           state.range(0),
+                           false,
+                           0,
+                           kNestedPayloadMessageType);
+}
+BENCHMARK_WITH_PERCENTILE(BM_NestedProtoPayloadFromJsonNonStreaming)
+    ->Arg(0) // flat JSON
+    ->Arg(1) // nested with 1 layer
+    ->Arg(8) // nested with 8 layers
+    ->Arg(32) // nested with 32 layers
+    ->Arg(64); // nested with 64 layers
+
+static void BM_NestedProtoPayloadFromJsonStreaming(::benchmark::State& state) {
+  BM_NestedPayloadFromJson(state,
+                           64,
+                           true,
+                           state.range(0),
+                           kNestedPayloadMessageType);
+}
+BENCHMARK_STREAMING_WITH_PERCENTILE(BM_NestedProtoPayloadFromJsonStreaming);
+
+static void BM_StructProtoPayloadFromJsonNonStreaming(::benchmark::State& state) {
+  BM_NestedPayloadFromJson(state,
+                           state.range(0),
+                           false,
+                           0,
+                           kStructPayloadMessageType);
+}
+BENCHMARK_WITH_PERCENTILE(BM_StructProtoPayloadFromJsonNonStreaming)
+    ->Arg(0) // flat JSON
+    ->Arg(1) // nested with 1 layer
+    ->Arg(8) // nested with 8 layers
+        // More than 32 layers would fail the parsing for struct proto.
+    ->Arg(32); // nested with 32 layers
+
+static void BM_StructProtoPayloadFromJsonStreaming(::benchmark::State& state) {
+  BM_NestedPayloadFromJson(state,
+                           kNumNestedLayersForStreaming,
+                           true,
+                           state.range(0),
+                           kStructPayloadMessageType);
+}
+BENCHMARK_STREAMING_WITH_PERCENTILE(BM_StructProtoPayloadFromJsonStreaming);
+
+//
+// Benchmark variable: Message chunk per message
+//
+// Helper function for benchmarking translation from segmented JSON input
+void BM_SegmentedStringPayloadFromJson(::benchmark::State& state,
+                                       uint64_t payload_length,
+                                       bool streaming,
+                                       uint64_t stream_size,
+                                       uint64_t chunk_per_msg) {
+  // We are using GetRandomAlphanumericString instead of GetRandomBytesString
+  // because JSON format reserves characters such as `"` and `\`.
+  // We could generate `"` and `\` and escape them, but for simplicity, we are
+  // only using alphanumeric characters.
+  // This would also be a more common for string proto.
+  auto msg =
+      absl::make_unique<std::string>(GetRandomAlphanumericString(payload_length));
+  auto json_msg =
```

Comment:
not need to use unique_ptr for json_msg and msg  just use std::string directly,   

actually, not need to use a variable for "msg",  just call the funtion directly in StrFormat().

Reply:
I used unique_ptr just to assign the string on the heap. Some test can use up 32 MiB of data, and heap is typically used to store a large chunk of data. All benchmarks are using heap to assign variables just to be consistent.

We may need `msg` variable as that was used later for the message parsing verification on line 460.

Reply:
I think even std::string is from stack,  its data() should be from heap.

Reply:
Sorry I just saw this. You're right :) I'm still learning C++ skills. Will change them to be on the stack.